File: implementations/cpp/order_statistic_trees.cpp
/**
 * @file order_statistic_trees.cpp
 * @brief Reference implementation of an Order-Statistic Tree (augmented balanced BST).
 *
 * Implements find_by_order(k) (select) and order_of_key(x) (rank) in O(log n) time
 * using an augmented Treap structure.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <random>
#include <cstdint>
#include <cassert>
#include <stdexcept>
#include <algorithm>

namespace dsa {

template <typename T>
class OrderStatisticTree {
private:
    struct Node {
        T key;
        uint32_t priority;
        size_t size;
        Node* left;
        Node* right;

        Node(const T& k, uint32_t p)
            : key(k), priority(p), size(1), left(nullptr), right(nullptr) {}
    };

    Node* root = nullptr;
    std::mt19937 rng;

    static size_t get_size(Node* n) {
        return n ? n->size : 0;
    }

    static void update_size(Node* n) {
        if (n) {
            n->size = 1 + get_size(n->left) + get_size(n->right);
        }
    }

    Node* rotate_right(Node* y) {
        Node* x = y->left;
        y->left = x->right;
        x->right = y;
        update_size(y);
        update_size(x);
        return x;
    }

    Node* rotate_left(Node* x) {
        Node* y = x->right;
        x->right = y->left;
        y->left = x;
        update_size(x);
        update_size(y);
        return y;
    }

    Node* insert_helper(Node* node, const T& key) {
        if (!node) {
            return new Node(key, rng());
        }
        if (key < node->key) {
            node->left = insert_helper(node->left, key);
            if (node->left->priority > node->priority) {
                node = rotate_right(node);
            }
        } else if (node->key < key) {
            node->right = insert_helper(node->right, key);
            if (node->right->priority > node->priority) {
                node = rotate_left(node);
            }
        }
        // If key == node->key, do not duplicate (set semantics)
        update_size(node);
        return node;
    }

    Node* erase_helper(Node* node, const T& key) {
        if (!node) return nullptr;

        if (key < node->key) {
            node->left = erase_helper(node->left, key);
        } else if (node->key < key) {
            node->right = erase_helper(node->right, key);
        } else {
            // Found node to delete
            if (!node->left && !node->right) {
                delete node;
                return nullptr;
            } else if (!node->left) {
                Node* temp = node->right;
                delete node;
                return temp;
            } else if (!node->right) {
                Node* temp = node->left;
                delete node;
                return temp;
            } else {
                // Both children exist: rotate child with higher priority to root
                if (node->left->priority > node->right->priority) {
                    node = rotate_right(node);
                    node->right = erase_helper(node->right, key);
                } else {
                    node = rotate_left(node);
                    node->left = erase_helper(node->left, key);
                }
            }
        }
        update_size(node);
        return node;
    }

    const T& select_helper(Node* node, size_t k) const {
        assert(node != nullptr);
        size_t left_size = get_size(node->left);
        if (k == left_size) {
            return node->key;
        } else if (k < left_size) {
            return select_helper(node->left, k);
        } else {
            return select_helper(node->right, k - left_size - 1);
        }
    }

    void destroy(Node* node) {
        if (node) {
            destroy(node->left);
            destroy(node->right);
            delete node;
        }
    }

public:
    OrderStatisticTree(uint32_t seed = 42) : rng(seed) {}

    ~OrderStatisticTree() {
        destroy(root);
    }

    // Disable copy for simplicity, allow move
    OrderStatisticTree(const OrderStatisticTree&) = delete;
    OrderStatisticTree& operator=(const OrderStatisticTree&) = delete;

    size_t size() const {
        return get_size(root);
    }

    bool empty() const {
        return root == nullptr;
    }

    void insert(const T& key) {
        root = insert_helper(root, key);
    }

    void erase(const T& key) {
        root = erase_helper(root, key);
    }

    /**
     * @brief Returns the k-th smallest element (0-indexed) in O(log n) time.
     */
    const T& find_by_order(size_t k) const {
        if (k >= size()) {
            throw std::out_of_range("Rank k exceeds tree size");
        }
        return select_helper(root, k);
    }

    /**
     * @brief Returns count of elements in the tree strictly smaller than key in O(log n) time.
     */
    size_t order_of_key(const T& key) const {
        size_t rank = 0;
        Node* curr = root;
        while (curr) {
            if (key <= curr->key) {
                curr = curr->left;
            } else {
                rank += get_size(curr->left) + 1;
                curr = curr->right;
            }
        }
        return rank;
    }
};

} // namespace dsa

int main() {
    std::cout << "Running Order-Statistic Tree C++17 unit tests..." << std::endl;

    dsa::OrderStatisticTree<int> ost(12345);

    // Test 1: Empty and size
    assert(ost.empty());
    assert(ost.size() == 0);

    // Test 2: Insert elements: {10, 20, 5, 15, 30}
    ost.insert(10);
    ost.insert(20);
    ost.insert(5);
    ost.insert(15);
    ost.insert(30);
    assert(ost.size() == 5);

    // Sorted keys should be: [5, 10, 15, 20, 30]
    // Test find_by_order (0-indexed select)
    assert(ost.find_by_order(0) == 5);
    assert(ost.find_by_order(1) == 10);
    assert(ost.find_by_order(2) == 15);
    assert(ost.find_by_order(3) == 20);
    assert(ost.find_by_order(4) == 30);

    // Test order_of_key (rank)
    assert(ost.order_of_key(5) == 0);
    assert(ost.order_of_key(10) == 1);
    assert(ost.order_of_key(12) == 2); // 5, 10 are smaller
    assert(ost.order_of_key(15) == 2);
    assert(ost.order_of_key(35) == 5); // all 5 are smaller

    // Test 3: Deletion
    ost.erase(15);
    assert(ost.size() == 4);
    // Remaining: [5, 10, 20, 30]
    assert(ost.find_by_order(2) == 20);
    assert(ost.order_of_key(20) == 2);

    // Test 4: Randomized stress verification against sorted vector oracle
    {
        dsa::OrderStatisticTree<int> stress_ost(999);
        std::vector<int> oracle;
        std::mt19937 gen(42);
        std::uniform_int_distribution<int> dist(1, 10000);

        for (int i = 0; i < 2000; ++i) {
            int val = dist(gen);
            if (std::find(oracle.begin(), oracle.end(), val) == oracle.end()) {
                stress_ost.insert(val);
                oracle.push_back(val);
                std::sort(oracle.begin(), oracle.end());
            }
        }

        assert(stress_ost.size() == oracle.size());
        for (size_t k = 0; k < oracle.size(); k += 50) {
            assert(stress_ost.find_by_order(k) == oracle[k]);
            assert(stress_ost.order_of_key(oracle[k]) == k);
        }
    }

    std::cout << "[PASS] All Order-Statistic Tree C++ unit tests passed." << std::endl;
    return 0;
}

File: implementations/python/order_statistic_trees.py
"""
Order-Statistic Tree (Augmented Balanced Binary Search Tree).

Implements find_by_order(k) (select) and order_of_key(x) (rank) in O(log n) time
using an augmented Treap data structure.
"""

import random
import unittest
from typing import Optional


class Node:
    """Treap node augmented with subtree size."""

    def __init__(self, key: int, priority: int):
        self.key = key
        self.priority = priority
        self.size = 1
        self.left: Optional["Node"] = None
        self.right: Optional["Node"] = None


class OrderStatisticTree:
    """Augmented Treap-based Order-Statistic Tree."""

    def __init__(self, seed: int = 42):
        self.root: Optional[Node] = None
        self.rng = random.Random(seed)

    @staticmethod
    def _get_size(node: Optional[Node]) -> int:
        return node.size if node else 0

    @classmethod
    def _update_size(cls, node: Optional[Node]) -> None:
        if node:
            node.size = 1 + cls._get_size(node.left) + cls._get_size(node.right)

    def _rotate_right(self, y: Node) -> Node:
        x = y.left
        assert x is not None
        y.left = x.right
        x.right = y
        self._update_size(y)
        self._update_size(x)
        return x

    def _rotate_left(self, x: Node) -> Node:
        y = x.right
        assert y is not None
        x.right = y.left
        y.left = x
        self._update_size(x)
        self._update_size(y)
        return y

    def _insert(self, node: Optional[Node], key: int) -> Node:
        if not node:
            return Node(key, self.rng.randint(1, 10**9))
        if key < node.key:
            node.left = self._insert(node.left, key)
            if node.left.priority > node.priority:
                node = self._rotate_right(node)
        elif key > node.key:
            node.right = self._insert(node.right, key)
            if node.right.priority > node.priority:
                node = self._rotate_left(node)
        self._update_size(node)
        return node

    def _erase(self, node: Optional[Node], key: int) -> Optional[Node]:
        if not node:
            return None
        if key < node.key:
            node.left = self._erase(node.left, key)
        elif key > node.key:
            node.right = self._erase(node.right, key)
        else:
            if not node.left and not node.right:
                return None
            elif not node.left:
                return node.right
            elif not node.right:
                return node.left
            else:
                if node.left.priority > node.right.priority:
                    node = self._rotate_right(node)
                    node.right = self._erase(node.right, key)
                else:
                    node = self._rotate_left(node)
                    node.left = self._erase(node.left, key)
        self._update_size(node)
        return node

    def insert(self, key: int) -> None:
        self.root = self._insert(self.root, key)

    def erase(self, key: int) -> None:
        self.root = self._erase(self.root, key)

    def size(self) -> int:
        return self._get_size(self.root)

    def find_by_order(self, k: int) -> int:
        """Returns the k-th smallest element (0-indexed) in O(log n)."""
        if k < 0 or k >= self.size():
            raise IndexError("Index k out of range")
        curr = self.root
        while curr:
            left_size = self._get_size(curr.left)
            if k == left_size:
                return curr.key
            elif k < left_size:
                curr = curr.left
            else:
                k -= left_size + 1
                curr = curr.right
        raise RuntimeError("Unreachable")

    def order_of_key(self, key: int) -> int:
        """Returns number of elements strictly smaller than key in O(log n)."""
        rank = 0
        curr = self.root
        while curr:
            if key <= curr.key:
                curr = curr.left
            else:
                rank += self._get_size(curr.left) + 1
                curr = curr.right
        return rank


class TestOrderStatisticTree(unittest.TestCase):
    def test_basic_operations(self):
        ost = OrderStatisticTree()
        for x in [10, 20, 5, 15, 30]:
            ost.insert(x)

        self.assertEqual(ost.size(), 5)
        # Sorted: 5, 10, 15, 20, 30
        self.assertEqual(ost.find_by_order(0), 5)
        self.assertEqual(ost.find_by_order(1), 10)
        self.assertEqual(ost.find_by_order(2), 15)
        self.assertEqual(ost.find_by_order(3), 20)
        self.assertEqual(ost.find_by_order(4), 30)

        self.assertEqual(ost.order_of_key(5), 0)
        self.assertEqual(ost.order_of_key(10), 1)
        self.assertEqual(ost.order_of_key(12), 2)
        self.assertEqual(ost.order_of_key(35), 5)

        ost.erase(15)
        self.assertEqual(ost.size(), 4)
        self.assertEqual(ost.find_by_order(2), 20)
        self.assertEqual(ost.order_of_key(20), 2)

    def test_stress_against_sorted_oracle(self):
        ost = OrderStatisticTree(seed=999)
        oracle = []
        rng = random.Random(123)

        for _ in range(1000):
            val = rng.randint(1, 5000)
            if val not in oracle:
                ost.insert(val)
                oracle.append(val)
                oracle.sort()

        self.assertEqual(ost.size(), len(oracle))
        for idx in range(0, len(oracle), 25):
            self.assertEqual(ost.find_by_order(idx), oracle[idx])
            self.assertEqual(ost.order_of_key(oracle[idx]), idx)


if __name__ == "__main__":
    unittest.main()
